Add e2e admission-controller role access tests

katyanna opened the pull request:
Extend test roles Admission Controller coverage for EKS.

| **User Role \\ Namespace** | **user namespace** | **visibility**     | **kube-system** |
|----------------------------|---------------------|---------------------|------------------|
| **admin**                  | ✅ Write Access     | ✅ Write Access     | ✅ Write Access  |
| **collaborator**           | ✅ Write Access     | ✅ Write Access     |  ❌ No Access    |
| **engineer**               | ✅ Write Access     | ❌ No Access        | ❌ No Access     |

katyanna committed:
Add e2e admission-controller tests for Collaborator Role

Signed-off-by: Katyanna Moura <amelie.kn@gmail.com>

Commit:
add more skeleton code to cover remaining test cases

linki commented on cluster/cluster.yaml:
```diff
@@ -366,6 +366,37 @@ Resources:
+            - '/{{`{{SessionName}}`}}'
+      KubernetesGroups:
+      - zalando:engineer
+      Type: "STANDARD"
```
This is just to get a token with `zalando:engineer` group.

linki commented on test/e2e/authorization.go:
```diff
@@ -663,6 +708,7 @@ var _ = g.Describe("Authorization via admission-controller [RBAC] [Zalando]", fu
 				framework.ExpectNoError(err)
 			})
 
+			// TODO: is that correct?? We use "readonly" as the role in the test case, why should this be able to delete something?
```
Ok, it's a bit annoying.

The tests are designed in this way: The user in the the test case is always an "admin" in terms of RBAC. This is to ensure that any "prevention" must come from admission-controller, rather than from RBAC. Unfortunately, that makes the expectations here sort of wrong (maybe).

linki replied:
So, the tests here are only for the admission-controller's response of the request. They don't represent the real response which is a combination of RBAC+admission-controller.

Commit:
update testcases for kube-systema and visibility namespace protection

linki commented on test/e2e/authorization.go:
```diff
@@ -642,6 +643,19 @@ var _ = g.Describe("Authorization via admission-controller [RBAC] [Zalando]", fu
 				framework.ExpectNoError(err, "failed to create pod: %s in namespace: %s", nonSystemResource.Name, nonSystemResource.Namespace)
 			})
 
+			// Not needed actually
```
I commented this out because the missing test cases were about the namespaces itself rather than resources within the namespace.

But we could also add this test case. We would just need to create a resource first, otherwise the delete call doesn't work correctly.

linki commented on test/e2e/authorization.go:
```diff
@@ -651,6 +665,41 @@ var _ = g.Describe("Authorization via admission-controller [RBAC] [Zalando]", fu
 				gomega.Expect(err).To(gomega.MatchError(gomega.ContainSubstring("write operations are forbidden")))
 			})
+
+			// Not needed actually
```
See similar comment above.

linki commented on test/e2e/authorization.go:
```diff
@@ -717,6 +767,72 @@ var _ = g.Describe("Authorization via admission-controller [RBAC] [Zalando]", fu
+
+			g.It("should deny deletion of kube-system namespace", func() {
+				err := client.CoreV1().Namespaces().Delete(context.Background(), "kube-system", metav1.DeleteOptions{DryRun: []string{"All"}})
+				gomega.Expect(err).To(gomega.MatchError(gomega.ContainSubstring("this namespace may not be deleted")))
```
This error message doesn't come from `admission-controller` but from Kubernetes itself: https://github.com/kubernetes/kubernetes/blob/v1.32.4/staging/src/k8s.io/apiserver/pkg/admission/plugin/namespace/lifecycle/admission.go#L80 (immortal namespace). That's why the error message is different than the one for `visibility`.

linki commented on cluster/manifests/02-visibility/01-namespace.yaml:
```diff
@@ -2,3 +2,5 @@ apiVersion: v1
 metadata:
   name: visibility
+  labels:
+    admission.zalando.org/infrastructure-component: "true"
```
This is what protects the `visibility` namespace. We could do the same for `kube-system` (to maybe get the same error message) but we don't have a YAML for that at the moment.

linki commented:
If we can we should also add a test case for this - currently untested - behaviour: https://github.com/zalando-incubator/kubernetes-on-aws/blob/6e651edb911e143051d5b0de6ac655a85b649015/cluster/manifests/02-admission-control/teapot.yaml#L578-L586

katyanna commented on test/e2e/authorization.go:
```diff
@@ -717,6 +767,72 @@ var _ = g.Describe("Authorization via admission-controller [RBAC] [Zalando]", fu
+				gomega.Expect(err).To(gomega.MatchError(gomega.ContainSubstring("this namespace may not be deleted")))
+			})
+
+			g.It("should deny deletion of visibility namespace", func() {
```
`collaborator` role should not be able to delete the `collaborator` namespace?

linki replied:
Yes.

But even if not there's no point in deleting this namespace.

katyanna committed:
cleanup test cases and naming

- Rename privileged to admin
- Rename unprivileged to read-only

Keeping closer correlation with the okta groups.

Stick with `delete` as the write action exercised, so we cover the destructive access
similarly to the other tests.

Signed-off-by: Katyanna Moura <amelie.kn@gmail.com>

katyanna commented on test/e2e/authorization.go:
```diff
@@ -651,6 +665,41 @@ var _ = g.Describe("Authorization via admission-controller [RBAC] [Zalando]", fu
+				framework.ExpectNoError(err)
+			})
+
+			// these tests are similar to the ones for unprivileged (let's think about that) (also see remarks of the first test in the block below)
```
Let's keep all access tested and not depend on tests for unprivileged even if this role is built upon it.

linki commented on test/e2e/authorization.go:
```diff
@@ -849,8 +969,13 @@ func getCollaboratorClient(cluster *types.Cluster, awsAccountID string) (*kubern
+
+// getReadOnlyClient returns a client with the `zalando:readonly` group.
+func getReadOnlyClient(cluster *types.Cluster, awsAccountID string) (*kubernetes.Clientset, error) {
 	return newClientWithRole(cluster, fmt.Sprintf("arn:aws:iam::%s:role/%s-e2e-eks-iam-test-unprivileged-role", awsAccountID, aws.ToString(cluster.Name)))
```
Let's also rename the role itself from unprivileged to read-only.

linki commented on test/e2e/authorization.go:
```diff
@@ -839,8 +959,8 @@ var _ = g.Describe("Authorization via admission-controller [RBAC] [Zalando]", fu
-func getPrivilegedClient(cluster *types.Cluster, awsAccountID string) (*kubernetes.Clientset, error) {
+// getAdminClient returns a client with the `zalando:administrator` group.
+func getAdminClient(cluster *types.Cluster, awsAccountID string) (*kubernetes.Clientset, error) {
 	return newClientWithRole(cluster, fmt.Sprintf("arn:aws:iam::%s:role/%s-e2e-eks-iam-test-privileged-role", awsAccountID, aws.ToString(cluster.Name)))
```
Let's also rename the role itself from privileged to admin/administrator.

linki commented on test/e2e/authorization.go:
```diff
@@ -711,19 +752,98 @@ var _ = g.Describe("Authorization via admission-controller [RBAC] [Zalando]", fu
+			g.It("should deny write access for system resources", func() {
+				err := client.RbacV1().ClusterRoles().Delete(context.Background(), systemResource.Name, metav1.DeleteOptions{DryRun: []string{"All"}})
+				gomega.Expect(err).To(gomega.MatchError(gomega.ContainSubstring("write operations are forbidden")))
+			})
```
Let's add the third test case from `collaborator` here as well to test that it's different, e.g.

```
g.It("should deny write access for collaborator resources", func() {
```

linki replied:
Unless that's not what we want. I'm not sure actually.

linki commented on test/e2e/authorization.go:
```diff
@@ -711,19 +752,98 @@ var _ = g.Describe("Authorization via admission-controller [RBAC] [Zalando]", fu
 			})
 
+			g.It("should allow write access for collaborator resources", func() {
+				err := client.RbacV1().ClusterRoles().Delete(context.Background(), collaboratorResource.Name, metav1.DeleteOptions{DryRun: []string{"All"}})
```
`collaboratorResource` is not initialized in the `BeforeEach`. But given that the outer context is "for global resources", it doesn't make sense to have collaborator resources here because we don't have this concept yet. When referring to the collaborator namespace itself, we can just use the string "visibility".

katyanna committed:
rename actual roles

Signed-off-by: Katyanna Moura <amelie.kn@gmail.com>

katyanna committed:
remove collaborator related test cases from global resources

Signed-off-by: Katyanna Moura <amelie.kn@gmail.com>

katyanna committed:
Remove read-only testcases from Admission Controller section

Admission Controller does not cover read-only permissions, that's
entirely on the RBAC.

Signed-off-by: Katyanna Moura <amelie.kn@gmail.com>

katyanna commented:
`read-only` role is not scoped, since it's entirely handled by RBAC.

demonCoder95 commented:
👍

katyanna commented:
:+1:
